common/include/pcl/pcl_macros.h

```diff
@@ -365,22 +365,20 @@ aligned_free (void* ptr)
  *
  * \ingroup common
  */
-#if __has_cpp_attribute(fallthrough) && !(defined(__clang__) && __cplusplus < 201703L)
+#if (__cplusplus >= 201703L) || (defined(_MSC_VER) && (_MSC_VER >= 1910) && (_MSVC_LANG >= 201703L))
   #define PCL_FALLTHROUGH [[fallthrough]];
 #elif defined(__clang__)
   #define PCL_FALLTHROUGH [[clang::fallthrough]];
-#elif defined(__GNUC__)
-  #if __GNUC__ >= 7
-    #define PCL_FALLTHROUGH [[gnu::fallthrough]];
-  #else
-    #define PCL_FALLTHROUGH ;
-  #endif
+#elif defined(__GNUC__) && (__GNUC__ >= 7)
+  #define PCL_FALLTHROUGH [[gnu::fallthrough]];
 #else
-  #define PCL_FALLTHROUGH ;
+  #define PCL_FALLTHROUGH
 #endif
 
-#if __has_cpp_attribute(nodiscard)
+#if (__cplusplus >= 201703L) || (defined(_MSC_VER) && (_MSC_VER >= 1911) && (_MSVC_LANG >= 201703L))
   #define PCL_NODISCARD [[nodiscard]]
+#elif defined(__GNUC__)
+  #define PCL_NODISCARD [[gnu::warn_unused_result]]
```

Comment:
clang also appears to have `[[clang::warn_unused_result]]`

http://releases.llvm.org/4.0.0/tools/clang/docs/AttributeReference.html#nodiscard-warn-unused-result-clang-warn-unused-result-gnu-warn-unused-result

Reply:
It is already [available](http://releases.llvm.org/3.9.0/tools/clang/docs/ReleaseNotes.html#id8) in 3.9. Only sad thing: I don't know which features were already present in  versions before `201703L`, as clang supports C++14 since 3.4, but it seems `[[fallthrough]]` & `[[nodiscard]]` were added in 3.9, so I hope before 3.9 the value of `__cplusplus` is not `201703L`.

Reply:
>  so I hope before 3.9 the value of `__cplusplus` is not `201703L`.

In theory it should reflect whatever is passed to `-std=c****`. AFAIK, only MSVC was not setting this value properly.

Reply:
Well sure, but GCC 5.2 reports with C++17 enabled `201500L`, during GCC 9.2 reports `201703L` (tested with online compiler on cppreference.com)